### PR TITLE
Feature todo app typescript

### PR DESCRIPTION
**Pull request by franbe7**
# :recycle: Refactor: ToDoApp Typescript

#### :link: Trello reference:

- [ToDoApp Typescript](https://trello.com/c/m3ZR2Ouc/35-todoapp-typescript)

---

#### :information_source: Description:

This PR contains  the whole refactor for typing the TodoApp with typescript

### CODE REVIEW

File: src/actions/ActionHandlers.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { TasksState, NewTaskState } from 'src/reducers/types'
+import { Actions } from 'src/actions/types'
+import { Task } from 'src/types/global'
+
+const _ = require('lodash')
+
+export const getTasks = (state: TasksState, action: Actions) => {
+  return Object.assign({}, state, {
```
Comment (nicoache1) [inline src/actions/ActionHandlers.ts]: Just a comment, you can use the spread `...` instead the `Object.assign`, just only a comment, you dont need to change this.

File: src/actions/ActionHandlers.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    tasks: _.get(action, 'payload.tasks', ''),
+  })
+}
+
+export const addTask = (state: TasksState, action: Actions) => {
+  return Object.assign({}, state, {
```
Comment (nicoache1) [inline src/actions/ActionHandlers.ts]: also another way to do this is
`return {
    ...state,
    tasks: [...state.tasks, _.get(action, 'payload.task', [])],
    error: {},
}`

File: src/actions/index.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  failureAction,
+} from 'src/actions/types'
+import { ActionCreator, Dispatch } from 'redux'
+import { Task } from 'src/types/global'
+
+export interface newTodo {
+  title: string
+}
+
+export const getTasks: any = () => {
```
Comment (nicoache1) [inline src/actions/index.ts]: Don't use any please, you can type this function

File: src/actions/index.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    try {
+      const res = await controller.getToDo()
+      dispatch(getTasksSuccess(res.data))
+    } catch (err) {
+      dispatch(getTasksFailure(err.message))
+    }
+  }
+}
+
+export const addTask: any = (title: string, description: string) => {
```
Comment (nicoache1) [inline src/actions/index.ts]: Same here

File: src/actions/index.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+export const addTaskSuccess: ActionCreator<taskAddedAction> = (task: Task) => ({
+  type: stringTypes.ADD_TASK_SUCCESS,
+  payload: {
+    task,
+  },
+})
+
+export const clearAllDone: any = (tasks: Task[]) => {
```
Comment (nicoache1) [inline src/actions/index.ts]: and here

File: src/actions/stringTypes.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-const types = {
+const stringTypes = {
```
Comment (nicoache1) [inline src/actions/stringTypes.ts:1]: please change the name here, we called this `actionTypes` and also, this needs to be a `ts` right?

File: src/containers/NewTask.tsx [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+export interface DispatchToProps {
+  addTask: (title: string, description: string) => void
+  onChangeForm: (title: string, description: string) => changeFormAction
+  resetForm: () => changeFormAction
+}
+
+export interface StateToProps {
+  title: ''
```
Comment (nicoache1) [inline src/containers/NewTask.tsx:106]: why you have this `''` as a type? if you want to type this a a string it needs to be `:string`, or maybe `?:string`

File: src/reducers/formNewTask.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  title: '',
+  description: '',
+}
+
+const formNewTaskReducer: Reducer<NewTaskState> = (
+  state = initial_state,
+  action: Actions,
+): NewTaskState => {
+  switch (action.type) {
+    case stringTypes.CHANGE_FORM:
```
Comment (nicoache1) [inline src/reducers/formNewTask.ts:17]: same here, `actionTypes`

File: src/reducers/tasks.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  getTasks,
+  addTask,
+  toggleDone,
+  clearAll,
+  failure,
+} from 'src/actions/ActionHandlers'
+
+const initial_state: TasksState = {
+  tasks: [],
+  error: '',
```
Comment (nicoache1) [inline src/reducers/tasks.ts]: maybe its better to use `undefined` what do you think?

File: src/scenes/home/Task.tsx [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+export interface Props {
+  url: string
+  title: string
+  description: string
+  completed: boolean
+  navigation: NavigationScreenProp<NavigationRoute>
+  toggle: (id: string) => void
+}
+
+function layoutTask(props: Props) {
```
Comment (nicoache1) [inline src/scenes/home/Task.tsx]: change this for
`export const LayoutTask:React.FunctionComponent<Props> = () => {`

1 - we use always arrow functions.
3 - just check this in every file

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,65 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es5" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */,
+    "module": "commonjs" /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */,
+    // "lib": [],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
```
Comment (nicoache1) [inline tsconfig.json]: maybe you can remove the commented code

File: src/scenes/home/ListTasks.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+import { Task } from 'src/scenes/home/Task'
+import { Task as typeTask } from 'src/types/global'
+
+export interface Props {
+  navigation: NavigationScreenProp<NavigationRoute>
+  tasks: typeTask[]
+}
+
+class ListTasks extends Component<Props> {
```
Comment (nicoache1) [inline src/scenes/home/ListTasks.tsx]: `export class ListTasks extends Component<Props> {`
don't use export default, always try to use export const

Comment (nicoache1) [inline src/scenes/home/ListTasks.tsx]: its neccesary to use a Component? you could solve it with just a `presentationalComponent`?

**Review by nicoache1 — CHANGES_REQUESTED** [review]

File: src/reducers/tasks.ts (hunk 9 above)
Comment (franbe7) [reply src/reducers/tasks.ts]: do you mean that i use undefined for `error` or for the initial state?

File: src/scenes/home/ListTasks.tsx (hunk 12 above)
Comment (franbe7) [reply src/scenes/home/ListTasks.tsx]: im agree with you

**Review by nicoache1 — APPROVED** [review]

**Pull request merged**